src/wattsi.pas

```diff
@@ -2207,22 +2209,36 @@ procedure Save(const Document: TDocument; const FileName: AnsiString; const InSp
                   Write(F, HighlighterOutputByJSONContents[CurrentHighlightedElementJSON])
                else
                begin
-                  if (AnsiContainsStr(ClassValue, 'idl')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/webidl?' + URLEncodedJSONContents)
-                  else
-                  if (AnsiContainsStr(ClassValue, 'css')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/css?' + URLEncodedJSONContents)
-                  else
-                  if (AnsiContainsStr(ClassValue, 'js')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/js?' + URLEncodedJSONContents)
-                  else
-                  if (AnsiContainsStr(ClassValue, 'abnf')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/abnf?' + URLEncodedJSONContents)
-                  else
-                  if (AnsiContainsStr(ClassValue, 'html')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/html?' + URLEncodedJSONContents);
-                  HighlighterOutputByJSONContents[CurrentHighlightedElementJSON] := HighlighterOutput;
-                  Write(F, Trim(HighlighterOutput));
+                  try
+                     HTTPClient := TFPHTTPClient.Create(nil);
+                     Ss := TStringStream.Create('');
+                     if (AnsiContainsStr(ClassValue, 'idl')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/webidl?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'css')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/css?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'js')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/js?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'abnf')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/abnf?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'html')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/html?' + URLEncodedJSONContents, Ss, [200,400]);
+                     HighlighterOutput := Ss.Datastring;
+                     Ss.Free;
+                     if HTTPClient.ResponseStatusCode = 400 then
+                     begin
+                        Writeln('Error: Highlighter failed.');
+                        Write(HighlighterOutput);
+                        Halt(1);
```

Comment:
We don't use Halt(1) anywhere else; instead we use the Fail() routine. Could we use that instead?

Reply:
So, that `Fail()` function we have is scoped to the `ProcessDocument()` function. And the highlighting happens in the `Save()` function, so we don’t have `Fail()` there unless we either copy the code from `ProcessDocument()` function, or else move `Fail()` up to (effectively) being a global function.

In writing this patch, I had considered making that `Fail()` a global so that we could (re)use it in the `Save()` function — but then I remembered that the reason we are using `Fail()` instead of `Halt()` to begin with in the `ProcessDocument()` code is that we need all of that code to first execute all of the way through before we exit. And the reason we need that is so that, after it runs and we finally do actually exit, we need to check the exit status to see if the cause was a parser error (in which case the exit status is 65 instead of 1), so that we can re-run to get the correct line numbers (for the actual source file instead of the pre-processed source we run it on the first time).

But the `Save()` code where the highlighter is called runs after the parser has finished (and so at that point we’re processing what’s effectively the DOM representation) — and therefore there’s no reason to do that `Fail()`-indirection dance, so we can safely just do a `Halt()` to catch fire and fail right away.

Reply:
Ah, this makes perfect sense then, thanks. I don't think we even need a comment or anything; I just missed that Fail() was scoped in that way.